src/lib/transaction_pool/transaction_pool.ml

```diff
@@ -28,19 +58,109 @@ module Make (Payment : sig
 
-  type t = {mutable pool: pool; log: Logger.t}
+  type t =
+    { mutable pool: pool
+    ; log: Logger.t
+    ; mutable diff_reader: unit Deferred.t Option.t }
 
-  let create ~parent_log =
-    { pool=
-        { heap= Fheap.create ~cmp:Payment.With_valid_signature.compare
-        ; set= Payment.With_valid_signature.Set.empty }
-    ; log= Logger.child parent_log __MODULE__ }
+  (* FIXME terrible hack *)
+  let remove_tx t tx =
+    let filter_out_tx =
+      (* The user commands in the breadcrumbs we get *have* been checked for
+         valid signatures, but we don't express that with types yet. *)
+      List.filter ~f:(fun tx' ->
+          User_command.compare tx (User_command.forget_check tx') <> 0 )
+    in
+    t.pool
+    <- { heap=
+           Fheap.to_list t.pool.heap |> filter_out_tx
+           |> Fheap.of_list ~cmp:User_command.With_valid_signature.compare
+       ; set=
+           User_command.With_valid_signature.Set.to_list t.pool.set
+           |> filter_out_tx |> User_command.With_valid_signature.Set.of_list }
+
+  let handle_diff t frontier
+      (diff_opt : Transition_frontier.Extensions.Best_tip_diff.view) =
+    match diff_opt with
+    | None ->
+        Logger.debug t.log "Got empty best tip diff" ;
+        Deferred.unit
+    | Some {old_best_tip; new_best_tip} ->
+        if
+          List.mem
+            (Transition_frontier.successors frontier old_best_tip)
+            new_best_tip ~equal:Breadcrumb.equal
+        then (
+          let new_user_commands = Breadcrumb.to_user_commands new_best_tip in
+          Logger.trace t.log
+            !"Diff: old: %{sexp:User_command.t list} new: \
+              %{sexp:User_command.t list}"
+            (Breadcrumb.to_user_commands old_best_tip)
+            new_user_commands ;
+          List.iter new_user_commands ~f:(fun tx -> remove_tx t tx) ;
+          printf !"removed txs\n%!" )
+        else
+          Logger.warn t.log
+            "Got a new best tip breadcrumb that wasn't a direct descendant of \
+             the old best tip. Handling this case is unimplemented." ;
+        Logger.trace t.log
+          !"Current transaction pool: \
+            %{sexp:User_command.With_valid_signature.t list}"
+          (Fheap.to_list t.pool.heap) ;
+        Deferred.unit
+
+  let create ~parent_log ~frontier_broadcast_pipe =
+    let t =
+      { pool=
+          { heap= Fheap.create ~cmp:User_command.With_valid_signature.compare
+          ; set= User_command.With_valid_signature.Set.empty }
+      ; log= Logger.child parent_log __MODULE__
+      ; diff_reader= None }
+    in
+    don't_wait_for
+      ( fst
+      @@ Broadcast_pipe.Reader.iter frontier_broadcast_pipe
+           ~f:(fun frontier_opt ->
+             match frontier_opt with
+             | None -> (
+                 Logger.debug t.log "no frontier" ;
+                 (* Sanity check: the view pipe should have been closed before
+                    the frontier was destroyed. *)
```

Comment:
I don't think this is hooked up properly yet. Look at the TODO on `close` in Transition-frontier.

Reply:
It gets closed in transition_router.ml:206. I'll go remove the comment.